src/main/core/worker.rs

```diff
@@ -141,6 +141,18 @@ impl Worker {
         .flatten()
     }
 
+    /// Run `f` with a reference to the global DNS.
+    ///
+    /// SAFETY: Assumes that the DNS has already been created and continues
+    /// to exist during the execution of `f`.
+    #[must_use]
+    pub fn with_dns<F, R>(f: F) -> R
```

Comment:
Thanks for the explanation! I do want to get the API design right so I appreciate you taking the time to explain the reasoning here :)

Reply:
I didn't spend too much time thinking about the API here because I know DNS will be changing, but I basically agree with you and it does seem worth it to get the `with_dns` api correct since we'll probably still keep that after rewriting DNS in rust. I'm going to hold off on the more detailed DNS bits because I do think those belong in a separate PR.

Reply:
I think this is UB since it allows two mutable references to the same memory:

```rust
Worker::with_dns(|dns_1| {
    Worker::with_dns(|dns_2| {
        // we have two mutable references to the DNS object (the nested `with_dns()` could be
        // in some other function where it's difficult to tell that you've invoked UB)
    });
});
```

Reply:
Oh crap, good catch! I made Jim's changes but in missed this very important sentence:

> We _would_ need to wrap it in something to ensure we don't create multiple mutable references though; probably making the dns field into a AtomicRefCell<SyncSendPointer<cshadow::DNS>>.

I submit a PR to fix this soon!